Merge bug mining into master

tomecho opened the pull request:
fixes #120, fixes #123, fixes #138  

# Outstanding Fixes Needed
- Update the minimize-patch and promote scripts.
- Update documentation standards in bug-mining scripts (incoming files are legacy bug-mining scripts with old headers).

# Major Changes Introduced
- All scripts that used to live only in the bug-mining branch are now ported to master.
- Bug-mining now uses a dedicated working directory and does not touch the core framework before a project and all its bugs are ready to be promoted to the main database.
- The `create-project.pl` script now takes care of the entire configuration and initialization -- no more manual steps required.
- The `merge-commit-db.pl` script's behavior for issuing bid (bug id's) changed: first respect existing bug ids then issue higher bids to newer bugs. Previously, a lower bid indicated a newer bug.
- The directories for a project's perl module and meta data can be globally configured in Defects4J to support bug mining in a dedicated working directory.
- `Project->checkout_vid` now has an additional parameter `is_bugmine`; if set to 1, the checkout will skip the processing of meta data, which may not exist during bug mining.
- New subroutines in each of the project modules (e.g., `framework/core/Project/Lang.pm`):
    * `determine_layout`: dynamically determine the directory layout and identify src and test directories for a checked-out revision. The results are cached in a file named *dir-layout.csv*.
    * `initialize_revision`: identify and log flaky or random tests -> bootstrapping routine, used during bug mining, to prepare a revision to be useful with the core framework.

rjust commented on framework/core/Project.pm:
```diff
@@ -370,55 +385,57 @@ sub checkout_vid {
     #       -> checkout and tag the pre-fix revision instead
-    $self->{_vcs}->export_diff($rev1, $rev2, $tmp_file);
-    $self->{_vcs}->apply_patch($work_dir, $tmp_file);
+    $self->export_diff($rev1, $rev2, $tmp_file, "src/");
```
What's the hard-coded "src/" argument?

tomecho replied:
I explicitly write src here to export the diff only from the src tree of the project, meaning `test/` and other directories will be excluded from any diff.

rjust replied:
The issue is that some projects store the sources in a differently named directory. For example, Chart's sources live in *source*. Also, isn't this diff supposed to include all the changes? This code computes and applies the diff from the post-fix revision to the pre-fix revision, which should be equivalent to checking out the pre-fix revision.

tomecho replied:
Okay, I will have to adjust that so it export the diff of the entire project.

rjust commented on framework/core/Project.pm:
```diff
@@ -276,32 +277,46 @@ sub sanity_check {
 
 Checks out the provided version id (C<vid>) to F<work_dir>, and tags the the buggy AND
 the fixed program version of this bug. Format of C<vid>: C<\d+[bf]>.
 The working directory (C<work_dir>) is optional, the default is C<prog_root>.
```
This comment needs to updated (`work_dir` should be `prog_root`, right?).

tomecho replied:
Yes.

rjust commented on framework/projects/defects4j.build.xml:
```diff
@@ -95,6 +95,17 @@ project-specific build file ("project_id"/"project_id".build.xml) for the
     </target>
 
+<!--
+    Light weight sanity check for bug mining
```
Can you add one sentence about what the sanity check does or is supposed to do in general? Seems like this is checking for the required properties.

rjust commented on framework/core/Project/Lang.pm:
```diff
@@ -42,18 +42,65 @@ use Vcs::Git;
     my $vcs  = Vcs::Git->new($PID,
                              "$REPO_DIR/$name.git",
-                             "$SCRIPT_DIR/projects/$PID/commit-db",
+                             (shift // "$SCRIPT_DIR/projects/$PID/commit-db"),
```
This subroutine now has two parameters. Can you make this more explicit by assigning them early on (`my (...) = @_`)? The call to `shift` at this location may cause trouble in the future.

Note that this comment also applies to the other project modules.

rjust commented:
@tomecho, overall this looks good. Can you please update the PR by pulling in the latest changes from master and by addressing the minor comments? Once all changes from master are pulled in, all bugs will be reproduced by travis, which would be a great sanity check.

tomecho commented:
Sounds good @rjust I should have that done by tomorrow evening.

rjust commented on framework/core/Project/Lang.pm:
```diff
@@ -42,18 +42,65 @@ use Vcs::Git;
     my $vcs  = Vcs::Git->new($PID,
                              "$REPO_DIR/$name.git",
-                             "$SCRIPT_DIR/projects/$PID/commit-db",
+                             ($commit_db // "$SCRIPT_DIR/projects/$PID/commit-db"),
```
Minor: can you move this up to line 50?

Also, shouldn't `$SCRIPT_DIR/projects` be `work_dir`?

rjust commented on framework/core/Project.pm:
```diff
@@ -1064,6 +1087,7 @@ sub _ant_call {
     @_ >= 2 or die $ARG_ERROR;
     my ($self, $target, $option_str, $log_file) =  @_;
     $option_str = "" unless defined $option_str;
+    # TODO why are we checking for this if we never use it
```
Line 1093 checks whether the build file exists. Do you mean that the code never uses the `_build_file` attribute?

tomecho replied:
It seems like we check if `$self->{_build_file}` exists but then when running ant we use `$D4J_BUILD_FILE`.  I think we had a discussion about this previously but I would like to double check, does ant import the project build file (`$self-{_build_file}`) from the master (`$D4J_BUILD_FILE`)?

rjust commented on framework/core/Project.pm:
```diff
@@ -982,6 +987,20 @@ sub contains_version_id {
 }
 
+=pod
+=item B<checkout_id> C<checkout_id(version [, work_dir])>
```
Can you change `version` to `revision_id`?

Also, `work_dir` should be `prog_root` in this routine.

tomecho replied:
`version` should really be a `vid` right?  VCS::checkout_vid accepts a `vid` and a `work_dir` (which should be named `prog_root`)

rjust commented on framework/core/Project.pm:
```diff
@@ -278,30 +279,31 @@ sub sanity_check {
 =pod
 
-  $project->checkout_vid(vid [, work_dir])
+  $project->checkout_vid(vid [, work_dir, is_bugmine])
```
The parameter `work_dir` should be named `prog_root`.

rjust commented:
Hey @tomecho, I added a few more comments -- mostly related to `prog_root` vs. `work_dir`.

Can you please address these comments? I can take it from there and address the remaining todos that you outlined.

Thanks,
René

Greg4cr commented:
download-issues.pl only asks for a project ID, but the URL also needs the organization name. For example, for project "jackson-dataformat-xml", you actually need the URL to be:

https://api.github.com/repos/FasterXML/jackson-dataformat-xml/issues?state=all&per_page=100&page=1

That is, it needs to include "FasterXML" (the organization) in addition to the project.

You can trick download-issues.pl into getting the right file by including the organization name in the project name:

./download-issues.pl github -p "FasterXML/jackson-dataformat-xml" -o working/issues/ | ./merge-issue-numbers.pl -f working/issues.tx

However, perhaps the organization should be included as a separate argument? Any thoughts?

In addition, the documentation for both this script and the overall bug mining process could make it clearer that this is not the project ID (JacksonXml), but the project name (jackson-dataformat-xml).

Greg4cr commented:
Typo in default project module:

sub initialize_revision {
    my ($self, $rev_id, $vid) = @_;
    $self->SUPER::initialize_revision($revision);
}

Line 2 should read:     my ($self, $revision, $vid) = @_;

Greg4cr commented:
The patch minimization stage currently fails. After the user updates the patch, d4j attempts to execute test cases to ensure that everything works as expected. However, the stage "update.all.tests" assumes the existence of a relevant tests file. During bug mining, these files do not exist.

tomecho commented:
> However, perhaps the organization should be included as a separate argument? Any thoughts?

This sounds like a good suggestion, I will make some changes to break it out in to a separate argument.  I should make it optional because certain issue trackers don't need the organization argument. Right?

Greg4cr commented:
Good point. GitHub needs it, but others might not.

On Mar 12, 2018 8:11 PM, "Tom Peck" <notifications@github.com> wrote:

> However, perhaps the organization should be included as a separate
> argument? Any thoughts?
>
> This sounds like a good suggestion, I will make some changes to break it
> out in to a separate argument. I should make it optional because certain
> issue trackers don't need the organization argument. Right?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/rjust/defects4j/pull/135#issuecomment-372503481>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB0-SkBycjMUi0m9IqWi8AW8SNLpH5t9ks5tdw6ZgaJpZM4RKmTT>
> .
>

tomecho commented:
Pushed some changes to keep the necessary organization param on download issues script for github, I made it possible to still supply it as `-p org/project` or in two args `-g org -p proj`

tomecho commented:
> In addition, the documentation for both this script and the overall bug mining process could make it clearer that this is not the project ID (JacksonXml), but the project name (jackson-dataformat-xml).

I added a line into the readme to explain this and also called the param tracker_project_id in the pod of download-issues.pl

Greg4cr commented:
The promote-to-db.pl script fails with error:
Invalid number of arguments! Use: create_project(project_id) at /home/greg/repos/defects4j-newbm/framework/core/Project.pm line 117.

The problem is the call to Project::create_project on line 115 of promote-to-db.pl.

Greg4cr commented:
To add to the above:

promote-to-db.pl fails because it calls a non-existent four-argument version of Project::create_project. The existing Project::create_project only takes the PID and assumes that the project you are adding is already part of the main D4J, and has no knowledge of the temporary directory.  The version that promote-to-db.pl calls supplies additional arguments that note the location of the working directory, the commit DB, and the build file.

@tomecho @rjust - Did either of you already write the code for this, and just forget to commit it?

tomecho commented:
There might have been something weird with a merge or something like that because I was able to use it on this branch around December 18th.  I'll poke around in the logs this weekend to see if anything got undone.  If not I'll just solve the issue.

 
 
  On Wed, Mar 21, 2018 at 3:45 PM, Gregory Gay<notifications@github.com> wrote:   
To add to the above:

promote-to-db.pl fails because it calls a non-existent four-argument version of Project::create_project. The existing Project::create_project only takes the PID and assumes that the project you are adding is already part of the main D4J, and has no knowledge of the temporary directory. The version that promote-to-db.pl calls supplies additional arguments that note the location of the working directory, the commit DB, and the build file.

@tomecho @rjust - Did either of you already write the code for this, and just forget to commit it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

rjust commented:
@Greg4cr @tomecho I made a number of (simplification) changes to the framework. I have not yet updated the promote script. In addition to updating the script to be compatible with the framework changes, we also need to add additional functionality for promoting all metadata (e.g., relevant tests) and for storing the issue tracker IDs in the commit-db.

tomecho commented:
As far as the issues with including the issue tracker name / id in the result db i have made some progress and i think it works, jsut need to do more testing.  The issue with promote-to-db is fixed.

tomecho commented:
@rjust I'm actually not entirely sure what needs to be done for adding the tracker id to the commit db and promoting metadata seeing as how the work dir has been merged back in with the regular framework directory. So far in terms of those work items I have just updated merge-commit-db to take a tracker id. i also added tracker id and tracker name to rev_list (although im not sure i was supposed to do that).

rjust commented on framework/bug-mining/README.md:
```diff
@@ -36,164 +37,213 @@ following files:
-    - github
-    - sourceforge (sf)
+    - [google-code](https://code.google.com/) ('google' for short)
+    - [jira](https://issues.apache.org/jira/)
```
Is apache's jira the default or the only one supported right now?

jose replied:
There is no need to define a default _tracker system_ as the `tracker-type` parameter of the `download-issues.pl` script isn't optional. And we currently support the following ones: google-code, jira, github, and sourceforge.

rjust commented on framework/bug-mining/promote-to-db.pl:
```diff
@@ -0,0 +1,273 @@
+
+The output directory as the Defects4J directory. Defaults to F<$PROJECTS_DIR>.
+
+=item B<-d C<output_db_dir>>
```
Do  we need this parameter? It seems like the database files are all temporary artifacts of the bug-mining process, and only the mined bugs and their meta data will be promoted to D4J proper.

rjust commented on framework/bug-mining/promote-to-db.pl:
```diff
@@ -0,0 +1,273 @@
+
+=item B<-o C<output_dir>>
+
+The output directory as the Defects4J directory. Defaults to F<$PROJECTS_DIR>.
```
Is `output_dir` essentially the target directory (in D4J proper)?

rjust commented on framework/bug-mining/promote-to-db.pl:
```diff
@@ -0,0 +1,273 @@
+=item B<-b C<bug_id>>
+
+Only analyze this bug id. The bug_id has to follow the format B<(\d+)(:(\d+))?>.
+Per default all bug ids, listed in the commit-db, are considered.
```
This should be "Only **promote** this bug is." and "By default, all bug ids, listed in the commit-db, are **promoted**".

rjust commented on framework/bug-mining/promote-to-db.pl:
```diff
@@ -0,0 +1,273 @@
+}
+
+#
+# Get version ids from TAB_TRIGGER
```
I think version id should be bug id.

Greg4cr commented:
Currently, 'sub _init_version' in initialize-revisions.pl does not work for projects that are subprojects of a greater project. That is, projects that are within a folder in the default work directory. For example, gson is actually in '$work_dir/gson'. 

This means I get a "Unsupported build system" error, even though Gson is actually a Maven project - there's not a 'pom.xml' file in the default work directory. I need to be able to point the script instead to '$work_dir/gson', which does have a 'pom.xml'. 

This same issue affects the functionality in the project file too. There, I can make it work by altering the project file to account for the directory shift. Of course, I could also do this in initialize-revisions.pl, but I wonder if we could support subprojects in a cleaner manner that does not require one-off edits to core bug mining files. Any thoughts?

Greg4cr commented on framework/bug-mining/download-issues.pl:
```diff
@@ -0,0 +1,290 @@
+
+=head1 NAME
+
+download-issues.pl -- Collect all issues from the project issue tracker.
```
Potential enhancement: 
For Github, the issue URL identifies if whether the matching number of an issue or a pull request. For example:
40,https://github.com/FasterXML/jackson-core/issues/40
39,https://github.com/FasterXML/jackson-core/pull/39
"40" represents an issue, while "39" represents a pull request. We could add an optional argument to filter this file for only issues, removing pull requests identified by the regex. This should be optional, rather than mandatory, as some pull requests also identify issues (GSON has examples of this).
